File: grokking_algorithms/ch1/binary_search.py
def binary_search(array, number):
    low = 0
    high = len(array) - 1

    while low <= high:
        mid = (low + high) // 2 # the guess index
        guess = array[mid]
        if guess == number:
            return mid
        if guess < number:
            low = mid + 1
        else:
            high = mid - 1
    return None

nums = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

assert binary_search(nums, 3) == 3, 'binary search for 3 failed'
assert binary_search(nums, 11) == None, 'binary search for 11 failed'
assert binary_search(nums, 9) == 9, 'binary search for 9 failed'

print('all ok')

File: grokking_algorithms/ch1/binarySearch.js
const assert = require('assert');

const arr = [-15, -5, -1, 0, 3, 5, 12, 59, 81, 333, 1539, 4964];
const binarySearch = (array, n) => {
  let min = 0;
  let max = array.length - 1;

  while (min <= max) {
    let guessIndex = Math.floor((max + min) / 2);
    let guess = array[guessIndex]; 
    if (guess === n) return guessIndex;
    if (guess < n) {
      min = guessIndex + 1;
    } else if (guess > n) {
      max = guessIndex - 1;
    }
  }
  return null;
}

const assertSearch = (n, i) => {
  assert.equal(binarySearch(arr, n), i, `finding ${n} has failed`);
}

const searches = [
  [-15, 0],      // find a small number
  [0, 3],        // find a med number
  [4964, 11],    // find a high number 
  [9999, null],  // non existing on the right
  [-9999, null], // non existing on the left
]

searches.forEach(search => assertSearch(...search));


